Ray_Serve/CV-Serve.ipynb

```diff
@@ -0,0 +1,1173 @@
+    "ingesting data, training models – including reinforcement learning\n",
+    "models – tuning those models and then serving them.\n",
+    "\n",
+    "<img src=\"https://technical-training-assets.s3.us-west-2.amazonaws.com/Introduction_to_Ray_AIR/e2e_air.png\" width=600 loading=\"lazy\"/>\n",
+    "\n",
+    "Key principles behind Ray and Ray AIR are\n",
+    "* Performance\n",
+    "* Developer experience and simplicity"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "dfc41d8f-4332-4359-b957-f26bc79de7a3",
+   "metadata": {},
+   "source": [
+    "# Ray Serve\n",
+    "\n",
+    "Serve is a microservices framework for serving ML – the model serving\n",
+    "component of Ray AIR.\n",
+    "\n",
+    "<img src='https://technical-training-assets.s3.us-west-2.amazonaws.com/Ray_Serve/serve_architecture.png' width=700/>\n",
+    "\n",
+    "# Deployments\n",
+    "\n",
+    "`Deployment` is the fundamental user-facing element of serve.\n",
+    "\n",
+    "<img src='https://technical-training-assets.s3.us-west-2.amazonaws.com/Ray_Serve/deployment.png' width=600/>\n",
+    "\n",
+    "## Our First Service\n",
+    "\n",
+    "Let’s jump right in and get something simple up and running on Ray\n",
+    "Serve."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "id": "48953d75-e69a-4fe3-adf8-079edc953683",
+   "metadata": {
+    "tags": []
+   },
+   "outputs": [],
+   "source": [
+    "@serve.deployment\n",
+    "class TextConverter:\n",
+    "    def convert(self, text):\n",
+    "        return \"***\" + str.upper(text) + \"***\""
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "id": "62641bcc-f2a0-48ee-ac1d-182688552af4",
+   "metadata": {
+    "tags": []
+   },
+   "outputs": [],
+   "source": [
+    "app_handle = serve.run(TextConverter.bind())"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "bcf7ce21-ee2a-48cb-8043-e986a4522e89",
+   "metadata": {},
+   "source": [
+    "## Key APIs and concepts\n",
+    "\n",
+    "`Deployment` represents a service and is created with the `@serve.deployment` decorator\n",
+    "* As end users, we don't instantiate `Deployment`s directly\n",
+    "* Ray will create them as actors, per our scaling requirements\n",
+    "\n",
+    "A __bound deployment__ is created with the `.bind` class method on the deployment class\n",
+    "* e.g., `TextConverter.bind(msg=\"Yes...\")` above creates a bound deployment\n",
+    "* `.bind` allows us to provide constructor params for the deployment class (the `msg` param above)\n",
+    "* bound deployments *can* be passed to other deployments via `.bind` -- this is one way to compose services\n",
+    "* We can pass a bound deployment to `serve.run(...)`\n",
+    "    * to start a service\n",
+    "    * to obtain a `ServeHandle`\n",
+    "\n",
+    "A `ServeHandle` can be used to invoke services through the Python API\n",
+    "* At runtime, services can call other services via serve handles\n",
```

Comment:
This is not exactly correct - deployments can call other deployments via serve handles.

Reply:
So I mostly refer to deployments as "services" because that's the concept they correspond closer to in general usage (cf. COM, CORBA, Java, OS, web, etc services: an combo of interface, calling convention, and the outer later of functionality) -- it's what most people understand -- whereas deployment has a another slightly different meaning in general usage (the placing/publishing of software and/or hardware into production).

I get the overlap and why Ray Serve service API elements might have been named Deployments, but it's quite confusing to folks who are experienced in the space but new to Ray. Specifically, one can create a Ray Serve Deployment that sort of paradoxically isn't a deployment because it's not deployed. Later we start a Serve application, which is a little-d deployment, but not a big-d Deployment.

And then there are bound deployments which sometimes automagically become serve handles but other times are not ... I'm all for a big renaming festival if we have a chance :)

Reply:
I think we should make a distinction between deployments and services. Service should be used to refer to the entire ray serve cluster that is deployed, which has routes (that go to applications or fastapi deployments etc.). The deployment piece is kind of intuitive, it's a "deployed" piece of code that is scaled independently. Multiple deployments can make up an application, multiple applications can make up a service. It is supposed to be distinctly different from typical webservices where the unit of deployment and service are equivalent. Which shows the uniqueness and power of ray serve